### PR TITLE
fixes #4998

### LINKED ISSUE #4998: Bug in IO module for BrainVision: Wrong conversion from Seconds to Hertz
In BrainVision Recorder *.vhdr files, the highpass filter settings are often listed as "low-cutoff in seconds". The relevant mne.io function translates this into Hz as such:

https://github.com/mne-tools/mne-python/blob/239c83dfe0686084d1c96c945579dc8cc10e895c/mne/io/brainvision/brainvision.py#L597-L598

With this method, the standard setting of `low-cutoff=10s` is translated to highpass=0.1Hz. However, the [BrainProducts website](http://www.brainproducts.com/productdetails.php?id=2) specifies 0.016Hz (see under "Lower cutoff frequency (high pass) / time constant").

The conversion from seconds to Hz is probably not implemented correctly in the mne.io function for brainvision. Or am I missing something?

Some references:
- [book excerpt with brief explanation and equation](https://books.google.se/books?id=z-PmHHsi4zYC&pg=PA41&lpg=PA41&dq=cut+off+frequency+conversion+hertz+and+seconds&source=bl&ots=g8Cf68JVOy&sig=CuDjd5p59iyyemfpuHu8AUlpnc0&hl=en&sa=X&ved=0ahUKEwi10t-z2trZAhVSsaQKHWh-DS84ChDoAQgmMAA#v=onepage&q=cut%20off%20frequency%20conversion%20hertz%20and%20seconds&f=false)

- [handy calculator and equations](http://www.sengpielaudio.com/calculator-timeconstant.htm)

### PR DESCRIPTION
**Pull request by sappelhoff**
fixes #4998 

Regarding the conversion from seconds to Hertz when reading filter information from a BrainVision Recorder *.vhdr file.

### CODE REVIEW

**Comment by larsoner** [comment]
@palday it looks like you worked on this, does it make sense to you?

@sappelhoff can you look to see if any of our test files use the time constant? If they do then we can update a test.

**Comment by sappelhoff** [comment]
@larsoner - that's right ... I completely forgot about the tests. Will check that tomorrow and update the PR if necessary!

File: mne/io/brainvision/brainvision.py [hunk 1]
```diff
@@ -595,7 +595,7 @@ def _get_vhdr_info(vhdr_fname, eog, misc, scale, montage):
             else:
                 info['highpass'] = float(highpass[0])
                 if hp_s:
-                    info['highpass'] = 1. / info['highpass']
+                    info['highpass'] = 1. / (2 * np.pi * info['highpass'])
```
Comment (agramfort) [inline mne/io/brainvision/brainvision.py]: can you add as comments the urls and explanation from the issue?
so in two year when we try to understand why there is a 2pi we know why.

also can you update a test so we don't break this in the future? thx

**Comment by palday** [comment]
@larsoner I'm happy to check things when all of @agramfort's requests have been implemented. (If you can be specific, e.g. section/page number, when pointing to the BV documentation, that would be particularly swell.) 

The general idea makes sense and I also have vague memory of time constant and analog filters .... 

This might be a "compatibility" / historical artifact from the way BrainVision records data. Their amplifiers can do analog (high-pass) filtering in hardware and BrainVision Recorder /  PyCorder can do digital filtering (both high- and low-pass) in software, so they might be aiming for a consistent interface across filter methods, although it makes it different across passband-edges.

**Comment by sappelhoff** [comment]
@palday I referenced a textbook (with page number) instead of the BrainProducts manual (there was no description of the conversion, just a mention of the 0.016Hz / 10s standard setting)

File: mne/io/brainvision/tests/test_brainvision.py [hunk 2]
```diff
@@ -240,8 +239,8 @@ def test_brainvision_data_lowpass_filters():
 
     assert_true(all(any([lp, hp]) for lp, hp in expected_warnings))
 
-    assert_equal(raw.info['highpass'], 0.1)
-    assert_equal(raw.info['lowpass'], 250.)
+    assert_equal(raw.info['highpass'], 1. / (2 * np.pi * 10))
+    assert_equal(raw.info['lowpass'], 1. / (2 * np.pi * 0.004))
```
Comment (palday) [inline mne/io/brainvision/tests/test_brainvision.py]: Rather than writing 1/250 as 0.004, I would put the 250 into the numerator to make it clear what the original Hz was. This also make a nice parallel between the Hertz and seconds code.

File: mne/io/brainvision/tests/test_brainvision.py [hunk 3]
```diff
@@ -214,16 +213,16 @@ def test_brainvision_data_lowpass_filters():
...
 
     # Homogeneous lowpass in seconds
     raw = _test_raw_reader(
         read_raw_brainvision, vhdr_fname=vhdr_lowpass_s_path,
         montage=montage, eog=eog, event_id=event_id)
 
-    assert_equal(raw.info['highpass'], 0.1)
-    assert_equal(raw.info['lowpass'], 250.)
+    assert_equal(raw.info['highpass'], 1. / (2 * np.pi * 10))
+    assert_equal(raw.info['lowpass'], 1. / (2 * np.pi * 0.004))
```
Comment (palday) [inline mne/io/brainvision/tests/test_brainvision.py]: Leave the "Hertz" obvious, see comment below

**Comment by palday** [comment]
@sappelhoff Two questions:

1. Are we sure this (analogue-style) conversion is also applied to the low-pass filters? The BV website has "1,000 / 250 Hz (switchable for resolutions 0.1 μV / 0.5 μV per bit)" and I'm not sure if that's a typo and should be "1,000s" or if it's reflecting the different resolutions (which would make less sense, I think, because it's 5x difference in resolution but only 4x in frequency).

2. Is the time-frequency conversion handled differently in "Amplifier Setup" and "Software Filters" (see my comment above and @larsoner's suspicion in #4998)? I fear that we'll have to dig through more BV technical documentation or maybe even contact Brain Products directly to get a straight answer to this question.

And a broader question, maybe @teonbrooks has some insight:

3. Are BV low-pass filters ever specified in time or only in frequency? I've never seen one specified in time and this could simplify certain parts of the code and remove some of the edge cases above. 

(Corollary: Are BV high-pass filters ever specified in frequency? If I recall correctly, the actual recording software shows the HP in s and the LP in Hz.)

If these questions are blowing up the PR, then my recommendation would be to fix the high-pass in Amplifier Setup, leave #4998 open until we figure out the answers and then do a follow-up PR.

**Comment by palday** [comment]
(Don't you just love it when a trivial change is full of arcane details?)

**Comment by sappelhoff** [comment]
@palday, thanks for the review ... those are some crucial details.

Regarding the conversion from seconds to Hertz, the formula: "1. / (2 * pi * t)", where t is the time constant should work for both high-pass and low-pass filters, according to page 40 to 42 in [this book](https://books.google.se/books?id=z-PmHHsi4zYC&pg=PA41&lpg=PA41&dq=cut+off+frequency+conversion+hertz+and+seconds&source=bl&ots=g8Cf68JVOy&sig=CuDjd5p59iyyemfpuHu8AUlpnc0&hl=en&sa=X&ved=0ahUKEwi10t-z2trZAhVSsaQKHWh-DS84ChDoAQgmMAA#v=onepage&q=cut%20off%20frequency%20conversion%20hertz%20and%20seconds&f=false). That being said, I don't know how BrainProducts implemented this conversion for their filters and it will probably be worth it, to ask the technical support about this (should I do this?). The questions would be:

1. Concerning hardware filters: Are the low-pass filters in .vhdr files *always* given in Hz?
    - This is @palday's question 3
    - If the answer is yes, we don't need to worry about conversion
    - If the answer is no, would the same conversion apply as for high-pass filters? ("1. / (2 * pi * t)")

2. Concerning software filters: Are the filters always specified in Hertz?
    - If yes, we don't need to worry about conversion
    - If the answer is no, would the same conversion apply as for high-pass filters? ("1. / (2 * pi * t)")

The outcome of these questions would also allow us to work on the cases in [test_brainvision.py](https://github.com/mne-tools/mne-python/blob/master/mne/io/brainvision/tests/test_brainvision.py) and cut out several cases that might never happen anyways.

Basically, the only thing we know for sure is, that the high-pass *hardware* filters are possibly always specified with a time constant t in seconds and that they can be converted to Hertz using "1. / (2 * pi * t)".

File: mne/io/brainvision/tests/test_brainvision.py (hunk 2 above)
Comment (sappelhoff) [reply mne/io/brainvision/tests/test_brainvision.py]: there is no "250" anywhere in the files --> it was computed using the "old formula" to convert seconds to Hertz: `1. / t`. In the files, t=0.004, so that's where the 250 come from. For the code, I suggest to leave it at 0.004 to keep the consistency of the formula `1. / (2 * pi * t)`

File: mne/io/brainvision/tests/test_brainvision.py (hunk 3 above)
Comment (sappelhoff) [reply mne/io/brainvision/tests/test_brainvision.py]: see above

**Comment by palday** [comment]
If you can get an answer direct from BP regarding these things, that would be optimal. Their outward facing staff might not know the answer, but they can surely asking their engineering colleagues. 

I would reword your question 2:

> How are your software filters specified, i.e. as time (s) or frequency (Hz)? If they are specified as time, does the conversion to frequency follow the "digital" convention f = 1 / t or the convention common in analog signal processing f = 1 / 2pi*t ? Based on the website, the latter convention seems to be used for amplifier settings, but it's unclear whether this is also the case for digital filters applied in software.

**Comment by sappelhoff** [comment]
I settled for:

>In each .vhdr file, there are specifications for hardware and software filters.
>
>Concerning the hardware filters, the high-pass filter is often given in the format "Low Cutoff [s]", which specifies a time constant to be converted into a frequency using the formula: f = 1. / (2 * pi * t) for a given time constant t. On the other hand, the low-pass filter is often given in the format "High Cutoff [Hz]".
>
>Question 1: Is the low-pass filter always given as a frequency [Hz]?
>
>Question 2: (only if answer to question 1 is "no") Should the low-pass filter sometimes be given as a time constant t [s], would it be converted to a frequency using f = 1. / 2 * pi * t?
>
>The third question is concerning the software filters:
>
>Question 3: How are your software filters specified,  as time (s) or frequency (Hz)? If they are specified as time, does the conversion to frequency follow the "digital" convention f = 1. / t or the convention common in analog signal processing f = 1. / 2 * pi * t ?

Let's hope for a nice answer :-)

**Comment by larsoner** [comment]
In my experience a few years ago the BV support team was quick to respond about technical filtering questions, so I'm optimistic. Thanks for sorting this out.

**Comment by sappelhoff** [comment]
We have an answer:

>The units are always specified in the header file. Therefore, regarding the hardware filters, while the High Cutoff is indicated in Hz (and it is always specified as [Hz]), the Low Cutoff is expressed in seconds ([s]), hence the value corresponds to the time constant. The relation between time constant in seconds and Cut off Frequency in Hz is the one you indicated in your Question 2, the same relation applies for the software filters.

**Comment by palday** [comment]
That's interesting. So it makes the s -> Hz conversion consistent.

The question remains whether we want to detect the unit: Native BV-produced files always have Hz for low-pass and s for high-pass, but the explicit unit coding does allow for other writers to use e.g. only Hz. (I'm curious if e.g. the EEGLAB bv-export code even bothers to export filter information and if so, in what format.) My tendency is to leave the code as is because it's already there, but it is a potential simplification to just support the BV-native output and throw an exception if other units are used.  

@larsoner  Thoughts?

**Comment by palday** [comment]
@sappelhoff While we wait for an answer, you can add this fix to the What's New. Pushing that change should also restart Travis so that you've got all your CI checks behind you if the answer comes back "leave it as is".

**Comment by sappelhoff** [comment]
sure, @palday ... would that be something like

    - Fix bug in :func:`mne.io.brainvision.brainvision._get_vhdr_info` to use the correct conversion for filters from time constant to frequency by `Stefan Appelhoff`_


in doc/whats_new.rst ?

**Comment by palday** [comment]
Yep, and add a target to your name link at the bottom, e.g.

```
.. _Stefan Appelhof: https://yourwebsite.edu
```

**Comment by larsoner** [comment]
I agree we should leave the unit-detection step in there. We don't know if other writers (or even BV in the future) will necessarily follow the low-as-s high-as-Hz convention consistently.

**Comment by palday** [comment]
Then asuming the CIs are happy, everything looks LGTM.

**Comment by larsoner** [comment]
Thanks @sappelhoff
